controllers/kafkatopic_controller.go

```diff
@@ -189,23 +165,36 @@ func (r *KafkaTopicReconciler) Reconcile(request reconcile.Request) (reconcile.R
 			return requeueWithError(reqLogger, "failed to create kafka topic", err)
 		}
 
-		if err = controllerutil.SetControllerReference(cluster, instance, r.Scheme); err != nil {
-			if !k8sutil.IsAlreadyOwnedError(err) {
-				return requeueWithError(reqLogger, "failed to set cluster controller reference on new KafkaTopic", err)
-			}
-		}
+	}
+
+	// set controller reference to parent cluster
+	if instance, err = r.ensureControllerReference(ctx, cluster, instance); err != nil {
+		return requeueWithError(reqLogger, "failed to ensure controller reference", err)
+	}
 
+	// ensure kafkaCluster label
+	if instance, err = r.ensureClusterLabel(ctx, cluster, instance); err != nil {
+		return requeueWithError(reqLogger, "failed to ensure kafkacluster label on topic", err)
 	}
 
 	// ensure a finalizer for cleanup on deletion
 	if !util.StringSliceContains(instance.GetFinalizers(), topicFinalizer) {
 		reqLogger.Info("Adding Finalizer for the KafkaTopic")
-		r.addFinalizer(instance)
+		instance.SetFinalizers(append(instance.GetFinalizers(), topicFinalizer))
 	}
 
 	// push any changes
-	if err = r.Client.Update(context.TODO(), instance); err != nil {
-		return requeueWithError(reqLogger, "failed to update KafkaTopic with controller reference", err)
+	// TODO (tinyzimmer): This is sometimes failing the first reconcile attempt
+	// still with an "object already modified" error. It's benign, but should dig
+	// deeper at some point.
+	if err = r.Client.Update(ctx, instance); err != nil {
+		return requeueWithError(reqLogger, "failed to update KafkaTopic", err)
+	}
+
+	// Fetch the updated object
+	instance = &v1alpha1.KafkaTopic{}
```

Comment:
This one is not required since update mutates the instance and updates the latest changes to it